lib/pos_staking_rewards.go

```diff
@@ -1,5 +1,271 @@
 package lib
 
-func (bav *UtxoView) DistributeStakingRewardsToSnapshotStakes(blockHeight uint64) error {
+import (
+	"math/big"
+
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+)
+
+func (bav *UtxoView) DistributeStakingRewardsToSnapshotStakes(blockHeight uint64, blockTimestampNanoSecs uint64) error {
+	// Check if we have switched from PoW to PoS yet. If we have not, then the PoS consensus
+	// has not started yet. We don't want to distribute any staking rewards until the PoS consensus begins.
+	if blockHeight < uint64(bav.Params.ForkHeights.ProofOfStake2ConsensusCutoverBlockHeight) {
+		return nil
+	}
+
+	// Retrieve the current EpochEntry.
+	currentEpochEntry, err := bav.GetCurrentEpochEntry()
+	if err != nil {
+		return errors.Wrapf(err, "DistributeStakingRewardsToSnapshotStakes: problem retrieving current EpochEntry: ")
+	}
+
+	// Check if the current epoch's timestamp is somehow greater than the block timestamp. This should never as long
+	// as timestamps are moving forward when connecting each block.
+	if currentEpochEntry.CreatedAtBlockTimestampNanoSecs >= blockTimestampNanoSecs {
+		return errors.Errorf("DistributeStakingRewardsToSnapshotStakes: current EpochEntry's CreatedAtBlockTimestampNanoSecs "+
```

Comment:
Should this wrap a `RuleError` instead? Here's the comment about the purpose of RuleError from the core codebase:
```
// RuleError is an error type that specifies an error occurred during
// block processing that is related to a consensus rule. By checking the
// type of the error the caller can determine that the error was due to
// a consensus rule and determine which consensus rule caused the issue.
```

Reply:
Good idea. This can be a RuleError since the blockTimestamp is passed from the block header that's currently being connected